lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1239 @@
+        params += helper.get_all_params(self.hidden_to_hidden, **tags)
+        return params
+
+    def get_output_shape_for(self, input_shape):
+        return input_shape[0], input_shape[1], self.num_units
+
+    def get_output_for(self, input, mask=None, **kwargs):
+        """
+        Compute this layer's output function given a symbolic input variable.
+
+        Parameters
+        ----------
+        input : theano.TensorType
+            Symbolic input variable.
+        mask : theano.TensorType
+            Theano variable denoting whether each time step in each
+            sequence in the batch is part of the sequence or not.  If ``None``,
+            then it is assumed that all sequences are of the same length.  If
+            not all sequences are of the same length, then it must be
+            supplied as a matrix of shape ``(n_batch, n_time_steps)`` where
+            ``mask[i, j] = 1`` when ``j <= (length of sequence i)`` and
+            ``mask[i, j] = 0`` when ``j > (length of sequence i)``.
+
+        Returns
+        -------
+        layer_output : theano.TensorType
+            Symbolic output variable.
+        """
+
+        # Treat all dimensions after the second as flattened feature dimensions
+        if input.ndim > 3:
+            input = input.reshape((input.shape[0], input.shape[1],
+                                   T.prod(input.shape[2:])))
+
+        # Input should be provided as (n_batch, n_time_steps, n_features)
+        # but scan requires the iterable dimension to be first
+        # So, we need to dimshuffle to (n_time_steps, n_batch, n_features)
+        input = input.dimshuffle(1, 0, 2)
+        seq_len, num_batch, _ = input.shape
+
+        if self.precompute_input:
+            # Because the input is given for all time steps, we can precompute
+            # the inputs to hidden before scanning. First we need to reshape
+            # from (seq_len, batch_size, num_inputs) to
+            # (seq_len*batch_size, num_inputs)
+            input = T.reshape(input,
+                              (seq_len*num_batch, -1))
+            input = helper.get_output(
+                self.input_to_hidden, input, **kwargs)
+
+            # Reshape back to (seq_len, batch_size, num_units)
+            input = T.reshape(input, (seq_len, num_batch, -1))
+
+        # We will always pass the hidden-to-hidden layer params to step
+        non_seqs = helper.get_all_params(self.hidden_to_hidden)
+        # When we are not precomputing the input, we also need to pass the
+        # input-to-hidden parameters to step
+        if not self.precompute_input:
+            non_seqs += helper.get_all_params(self.input_to_hidden)
+
+        # Create single recurrent computation step function
+        def step(input_n, hid_previous, *args):
+            # Compute the hidden-to-hidden activation
+            hid_pre = helper.get_output(self.hidden_to_hidden, hid_previous)
```

Comment:
Yeah, all I can say is that when I tested it it didn't hurt performance, you (or @skaae or someone else) are welcome to re-test on a different system/different use-cases!  I think you may be right that `non_seqs` is cruft now, although I am wondering if I left that in there because `scan` wants to know what the non-sequence shared variables are, not because you need to retrieve them from the arguments and use them (as they are within the function's scope).